### PR TITLE
add support for either-or versions to waitForServices

### PR DESCRIPTION
**Pull request by OutdatedVersion**
## :memo: Description

Add support for a dependency formatted as `{ name: 'z', versions: [1, 2] }` - either `v1.z` or `v2.z` will satisfy this.

### :dart: Relevant issues

#1026 

### :gem: Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## :checkered_flag: Checklist:

- [ ] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] **I have added tests that prove my fix is effective or that my feature works**
- [x] **New and existing unit tests pass locally with my changes**
- [ ] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: src/service-broker.js [hunk 1]
```diff
@@ -986,36 +986,85 @@ class ServiceBroker {
 	waitForServices(serviceNames, timeout, interval, logger = this.logger) {
 		if (!Array.isArray(serviceNames)) serviceNames = [serviceNames];
 
-		serviceNames = _.uniq(
+		class ServiceToken {
```
Comment (icebob) [inline src/service-broker.js:989]: I feel the PR is a little complicated. Please try to solve it without sub-classes. Regarding performance, it's not optimal to create sub-classes and instances of classes in a method.

File: src/service-broker.js [hunk 2]
```diff
@@ -986,36 +986,85 @@ class ServiceBroker {
...
+			constructor(name, version) {
+				this.name = name;
+				this.version = version;
+			}
+
+			get fullName() {
+				if (this.version) {
+					return `v${this.version}.${this.name}`;
```
Comment (icebob) [inline src/service-broker.js:997]: It's not that simple. Instead use `Service.getVersionedFullName`

File: test/unit/service-broker.spec.js [hunk 3]
```diff
@@ -1709,7 +1709,7 @@ describe("Test broker.waitForServices", () => {
 					services: ["posts"],
 					statuses: [{ name: "posts", available: true }]
 				});
-				expect(broker.registry.hasService).toHaveBeenCalledTimes(6);
+				expect(broker.registry.hasService).toHaveBeenCalledTimes(12);
```
Comment (icebob) [inline test/unit/service-broker.spec.js:1712]: Why is duplicated the number of calls?

**Review by icebob — CHANGES_REQUESTED** [review]

**Comment by icebob** [comment]
Are you going to work on it?

File: src/service-broker.js (hunk 1 above)
Comment (OutdatedVersion) [reply src/service-broker.js:989]: Class based was the cleanest/simplest solution I found working on this - I toyed with a pure object/function but that got out of hand imo. Do you have any suggestions for a different approach? Thanks for your time here

Comment (OutdatedVersion) [reply src/service-broker.js:989]: @icebob Would you be comfortable moving forward with this approach if we found a better place to store these classes?

**Comment by OutdatedVersion** [comment]
@icebob This has been reworked, if you have a moment to look. Thanks!

**Review by icebob — APPROVED** [review]
I like it, thanks!

**Pull request merged**